### PR TITLE
prior box operator for ssd

### LINKED ISSUE #6015: Port PriorBox Operator.
- [PriorBox.cpp](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/layers/PriorBox.cpp)
- [priorbox api](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/trainer_config_helpers/layers.py#L1098)
- Used in SSD detection algorithm.

### PR DESCRIPTION
**Pull request by wanghaox**
resolve #6015

### CODE REVIEW

File: paddle/operators/prior_box_op.cc [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+namespace paddle {
+namespace operators {
+
+class PriorBoxOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("Input"),
+                   "Input(X) of SequenceSliceOp should not be null.");
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: Please check this comment.

File: paddle/operators/prior_box_op.cc [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    PADDLE_ENFORCE(ctx->HasInput("Image"),
+                   "Input(Offset) of SequenceSliceOp should not be null.");
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: Same as above.

I think we only use the shape of image. Is it necessary to pass the full image ?

File: paddle/operators/prior_box_op.cc [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+    auto image_dims = ctx->GetInputDim("Image");
+    auto input_dims = ctx->GetInputDim("Input");
+    PADDLE_ENFORCE(image_dims.size() == 4,
+                   "The format of input tensor is NCHW.");
+
+    auto min_sizes = ctx->Attrs().Get<std::vector<int>>("min_sizes");
+    auto max_sizes = ctx->Attrs().Get<std::vector<int>>("max_sizes");
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: Any possible to make 'max_sizes' optional?

File: paddle/operators/prior_box_op.cc [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    auto variances = ctx->Attrs().Get<std::vector<float>>("variances");
+    auto input_aspect_ratio =
+        ctx->Attrs().Get<std::vector<float>>("aspect_ratios");
+    bool flip = ctx->Attrs().Get<bool>("flip");
+
+    PADDLE_ENFORCE_GT(min_sizes.size(), 0, "must provide min_size.");
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: Please refine this comment and make sure the first character is upper-case.
I think 'Size of min_size must be at least 1.' is more accurate.

File: paddle/operators/prior_box_op.cc [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    const int img_w = ctx->Attrs().Get<int>("img_w");
+    PADDLE_ENFORCE_GT(img_w, 0, "img_w should be larger than 0.");
+
+    const float step_h = ctx->Attrs().Get<float>("step_h");
+    PADDLE_ENFORCE_GT(step_h, 0.0, "step_h should be larger than 0.");
+    const float step_w = ctx->Attrs().Get<float>("step_w");
+    PADDLE_ENFORCE_GT(step_w, 0.0, "step_w should be larger than 0.");
+
+    const int layer_height = input_dims[3];
+    const int layer_width = input_dims[2];
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: The shape of input is NCHW or NCWH?

File: paddle/operators/prior_box_op.cc [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    // Since all images in a batch has same height and width, we only need to
+    // generate one set of priors which can be shared across all images.
+    dim_vec[0] = 1;
+    // 2 channels. First channel stores the mean of each prior coordinate.
+    // Second channel stores the variance of each prior coordinate.
+    dim_vec[1] = 2;
+    dim_vec[2] = layer_width * layer_height * num_priors * 4;
+    PADDLE_ENFORCE_GT(dim_vec[2], 0,
+                      "output_dim[2] must larger than 0."
+                      "check your data dims");
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: If possible, please find the illegal input. For example, layer_width is illegal or layer_height is illegal etc.

File: paddle/operators/prior_box_op.cc [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+  }
+};
+
+class PriorBoxOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  PriorBoxOpMaker(framework::OpProto* proto,
+                  framework::OpAttrChecker* op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("Input",
+             "(Tensor), "
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: Please give type info like default Tensor\<float\>.

File: paddle/operators/prior_box_op.cc [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+             "the input feature data of PriorBoxOp.");
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: Lack shape information and no need to start in a newline.

File: paddle/operators/prior_box_op.h [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+#include "paddle/operators/math/math_function.h"
+// #include "paddle/operators/strided_memcpy.h"
+
+namespace paddle {
+namespace operators {
+
+inline void expand_aspect_ratios(const std::vector<float> input_aspect_ratior,
+                                 bool flip,
+                                 std::vector<float>& output_aspect_ratior) {
+  constexpr float eps = 1e-6;
```
Comment (sweetsky0901) [inline paddle/operators/prior_box_op.h]: eps建议epsilon，在重构后的版本中看到好像都用这个词。
另外这个建议做成参数

File: paddle/operators/prior_box_op.cc [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("Input"),
+                   "Input(X) of PriorBoxOp should not be null.");
+    PADDLE_ENFORCE(ctx->HasInput("Image"),
+                   "Input(Offset) of PriorBoxOp should not be null.");
+
+    auto image_dims = ctx->GetInputDim("Image");
+    auto input_dims = ctx->GetInputDim("Input");
+    PADDLE_ENFORCE(image_dims.size() == 4,
+                   "The format of input tensor is NCHW.");
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.cc]: need to declare which input in error message.

File: paddle/operators/prior_box_op.cc [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+    PADDLE_ENFORCE(input_dims.size() == 4,
+                   "The format of input tensor is NCHW.");
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.cc]: need to declare which input in error message.

File: paddle/operators/prior_box_op.cc [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+    PADDLE_ENFORCE_LT(input_dims[2], image_dims[2],
+                      "The height of input must smaller than image.");
+
+    PADDLE_ENFORCE_LT(input_dims[3], image_dims[3],
+                      "The width of input must smaller than image.");
+
+    auto min_sizes = ctx->Attrs().Get<std::vector<int>>("min_sizes");
+    auto max_sizes = ctx->Attrs().Get<std::vector<int>>("max_sizes");
+    auto variances = ctx->Attrs().Get<std::vector<float>>("variances");
+    auto input_aspect_ratio =
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.cc]: input_aspect_ratio -> aspect_ratio

File: paddle/operators/prior_box_op.h [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+limitations under the License. */
+
+#pragma once
+#include "paddle/framework/op_registry.h"
+#include "paddle/operators/math/math_function.h"
+
+namespace paddle {
+namespace operators {
+
+inline void expand_aspect_ratios(const std::vector<float> input_aspect_ratior,
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.h]: Follow google code style  https://google.github.io/styleguide/cppguide.html#Function_Names

Comment (qingqing01) [reply paddle/operators/prior_box_op.h]:
```c++
const std::vector<float>& input_aspect_ratior
```

File: paddle/operators/prior_box_op.cc [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    if (variances.size() > 1) {
+      PADDLE_ENFORCE_EQ(variances.size(), 4,
+                        "Must and only provide 4 variance.");
+      for (size_t i = 0; i < variances.size(); ++i) {
+        PADDLE_ENFORCE_GT(variances[i], 0.0,
+                          "variance[%d] must be greater than 0.", i);
+      }
+    } else if (variances.size() == 1) {
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.cc]: For simply, we can only support `variances.size() == 4` as the same as V2 API. This  is compatible with `variances.size() == 1`, for this case, the user can pass 4 same value.

File: paddle/operators/prior_box_op.cc [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+    const float step_w = ctx->Attrs().Get<float>("step_w");
+    PADDLE_ENFORCE_GT(step_w, 0.0, "step_w should be larger than 0.");
+
+    const int layer_height = input_dims[2];
+    const int layer_width = input_dims[3];
+
+    std::vector<int64_t> dim_vec(5);
+    dim_vec[0] = 2;
+    dim_vec[1] = layer_height;
+    dim_vec[2] = layer_width;
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.cc]: remove line 93 and line 94, 

```c++
dim_vec[1] = input_dims[2];
dim_vec[2] = input_dims[3];
```

File: paddle/operators/prior_box_op.cc [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+    dim_vec[3] = num_priors;
+    dim_vec[4] = 4;
+    auto output_dim = framework::make_ddim(dim_vec);
+    ctx->SetOutputDim("Out", output_dim);
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.cc]: remove line 102

```c++
ctx->SetOutputDim("Out", dim_vec);
```

File: paddle/operators/prior_box_op.h [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+    if (img_h == 0 || img_w == 0) {
+      img_width = image->dims()[3];
+      img_height = image->dims()[2];
+    } else {
+      img_width = img_w;
+      img_height = img_h;
+    }
+
+    const int layer_width = input->dims()[3];
+    const int layer_height = input->dims()[2];
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.h]: `int64_t`, since the type in `input->dims()` is `int64_t`

File: paddle/operators/prior_box_op.h [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+    T* output_data = nullptr;
+    framework::Tensor output_cpu;
+    framework::Tensor* output_tensor;
+    out->mutable_data<T>(ctx.GetPlace());
+    if (platform::is_gpu_place(ctx.GetPlace())) {
+      output_cpu.mutable_data<T>(out->dims(), platform::CPUPlace());
+      output_tensor = &output_cpu;
+    } else {
+      output_tensor = out;
+    }
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.h]: Now support multi-device, there is no need to copy tensor from GPU to CPU.

File: paddle/operators/prior_box_op.h [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+      for (int h = 0; h < layer_height; ++h) {
+        for (int w = 0; w < layer_width; ++w) {
+          for (int i = 0; i < num_priors; ++i) {
+            for (int j = 0; j < 4; ++j) {
+              e_out(0, h, w, i, j) =
+                  std::min<T>(std::max<T>(e_out(0, h, w, i, j), 0.), 1.);
+            }
+          }
+        }
+      }
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.h]: 这里的clip看下能不能调用Eigen的函数，而不是4层for循环。

File: paddle/operators/prior_box_op.h [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+      for (int h = 0; h < layer_height; ++h) {
+        for (int w = 0; w < layer_width; ++w) {
+          for (int i = 0; i < num_priors; ++i) {
+            for (int j = 0; j < 4; ++j) {
+              e_out(1, h, w, i, j) = variances[j];
+            }
+          }
+        }
+      }
+    }
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.h]: 同上，这种赋值，看看下能不能调用Eigen的函数，而不是4层for循环。

File: python/paddle/v2/fluid/tests/test_prior_box_op.py [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+                                max(output[0, h, w, i, j], 0), 1)
+        # set the variance.
+        for h in range(self.layer_h):
+            for w in range(self.layer_w):
+                for i in range(self.num_priors):
+                    for j in range(4):
+                        if len(self.variances) == 1:
+                            output[1, h, w, i, j] = self.variances[0]
+                        else:
+                            output[1, h, w, i, j] = self.variances[j]
```
Comment (qingqing01) [inline python/paddle/v2/fluid/tests/test_prior_box_op.py]: Python单测需要改进，尽量向量操作，不用这么多层的for循环。

File: paddle/operators/prior_box_op.h (hunk 19 above)
Comment (qingqing01) [reply paddle/operators/prior_box_op.h]: 可以参考 https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/compare_op.h#L72 ，用` platform::Transform` 也行。

File: paddle/operators/prior_box_op.cc [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+                   "Prior boxes step across height, 0 for auto calculation.")
+        .SetDefault(0.0);
+    AddAttr<float>("offset",
+                   "(float) "
+                   "Prior boxes center offset.")
+        .SetDefault(0.5);
+    AddComment(R"DOC(
+Prior box operator
+Generate prior boxes for SSD(Single Shot MultiBox Detector) algorithm.
+Please get more information from the following papers:
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply paddle/operators/prior_box_op.cc]: I cannot see any information in this comments.

File: paddle/operators/prior_box_op.h (hunk 20 above)
Comment (qingqing01) [reply paddle/operators/prior_box_op.h]: 这里用broadcast即可~

File: paddle/operators/prior_box_op.h (hunk 9 above)
Comment (wanghaox) [reply paddle/operators/prior_box_op.h]: done

File: paddle/operators/prior_box_op.cc [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+                        "The length of min_size and max_size must be equal.");
+      for (size_t i = 0; i < min_sizes.size(); ++i) {
+        PADDLE_ENFORCE_GT(max_sizes[i], min_sizes[i],
+                          "max_size[%d] must be greater than min_size[%d].", i,
+                          i);
+        num_priors += 1;
+      }
+    }
+
+    if (variances.size() > 1) {
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.cc]: It seems line 68 should be removed.

File: paddle/operators/prior_box_op.h [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+
+    boxes->mutable_data<T>(ctx.GetPlace());
+    vars->mutable_data<T>(ctx.GetPlace());
+
+    auto e_boxes = framework::EigenTensor<T, 4>::From(*boxes);
+    for (int h = 0; h < layer_height; ++h) {
+      for (int w = 0; w < layer_width; ++w) {
+        float center_x = (w + offset) * step_width;
+        float center_y = (h + offset) * step_height;
+        float box_width, box_height;
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.h]: > template <typename Place, typename T>

If T is double, here also should be double.

File: python/paddle/v2/fluid/tests/test_prior_box_op.py [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+                        center_x - box_width / 2.) / self.image_w
+                    # ymin
+                    out_boxes[h, w, idx, 1] = (
+                        center_y - box_height / 2.) / self.image_h
+                    # xmax
+                    out_boxes[h, w, idx, 2] = (
+                        center_x + box_width / 2.) / self.image_w
+                    # ymax
+                    out_boxes[h, w, idx, 3] = (
+                        center_y + box_height / 2.) / self.image_h
```
Comment (qingqing01) [inline python/paddle/v2/fluid/tests/test_prior_box_op.py]: 这块代码可以简化下， 

```python
c_x = (w + self.offset) * self.step_w
c_y = (h + self.offset) * self.step_h
# ...
c_w = c_h = min_size/2.
out_boxes[h, w, idx, :] = [(c_x - c_w)/self.image_w, (c_y - c_h)/self.image_h, ..., ...]  
# ...
```
下面两处计算相同，可以使代码更短一些。

File: paddle/operators/prior_box_op.cc [hunk 26]
```diff
@@ -0,0 +1,161 @@
...
+namespace paddle {
+namespace operators {
+
+class PriorBoxOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("Input"),
+                   "Input(X) of PriorBoxOp should not be null.");
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: Not `Input(X)`, should be `Input(Input)`.

File: paddle/operators/prior_box_op.cc [hunk 27]
```diff
@@ -0,0 +1,161 @@
...
+    PADDLE_ENFORCE(ctx->HasInput("Image"),
+                   "Input(Offset) of PriorBoxOp should not be null.");
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: Input(Offset) --> Input(Image)

File: paddle/operators/prior_box_op.cc [hunk 28]
```diff
@@ -0,0 +1,161 @@
...
+
+    auto image_dims = ctx->GetInputDim("Image");
+    auto input_dims = ctx->GetInputDim("Input");
+    PADDLE_ENFORCE(image_dims.size() == 4, "The format of image is NCHW.");
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: I think `The layout of data is NCHW` is better.

File: paddle/operators/prior_box_op.cc [hunk 29]
```diff
@@ -0,0 +1,161 @@
...
+                      "The width of input must smaller than image.");
+
+    auto min_sizes = ctx->Attrs().Get<std::vector<int>>("min_sizes");
+    auto max_sizes = ctx->Attrs().Get<std::vector<int>>("max_sizes");
+    auto variances = ctx->Attrs().Get<std::vector<float>>("variances");
+    auto aspect_ratios = ctx->Attrs().Get<std::vector<float>>("aspect_ratios");
+    bool flip = ctx->Attrs().Get<bool>("flip");
+
+    PADDLE_ENFORCE_GT(min_sizes.size(), 0,
+                      "Size of min_size must be at least 1.");
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: min_size --> min_sizes

File: paddle/operators/prior_box_op.cc [hunk 30]
```diff
@@ -0,0 +1,161 @@
...
+      PADDLE_ENFORCE_GT(min_sizes[i], 0, "min_sizes[%d] must be positive.", i);
+    }
+
+    std::vector<float> aspect_ratios_vec;
+    ExpandAspectRatios(aspect_ratios, flip, aspect_ratios_vec);
+
+    int num_priors = aspect_ratios_vec.size() * min_sizes.size();
+    if (max_sizes.size() > 0) {
+      PADDLE_ENFORCE_EQ(max_sizes.size(), min_sizes.size(),
+                        "The length of min_size and max_size must be equal.");
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: length --> number

File: paddle/operators/prior_box_op.cc [hunk 31]
```diff
@@ -0,0 +1,161 @@
...
+
+    PADDLE_ENFORCE_LT(input_dims[2], image_dims[2],
+                      "The height of input must smaller than image.");
+
+    PADDLE_ENFORCE_LT(input_dims[3], image_dims[3],
+                      "The width of input must smaller than image.");
+
+    auto min_sizes = ctx->Attrs().Get<std::vector<int>>("min_sizes");
+    auto max_sizes = ctx->Attrs().Get<std::vector<int>>("max_sizes");
+    auto variances = ctx->Attrs().Get<std::vector<float>>("variances");
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.cc]: Should be `variances` optional?

File: paddle/operators/prior_box_op.cc (hunk 23 above)
Comment (wanghaox) [reply paddle/operators/prior_box_op.cc]: done

File: paddle/operators/prior_box_op.h [hunk 32]
```diff
@@ -0,0 +1,184 @@
...
+
+    if (clip) {
+      platform::Transform<platform::CPUDeviceContext> trans;
+      ClipFunctor<T> clip_func;
+      trans(ctx.template device_context<platform::CPUDeviceContext>(),
+            boxes->data<T>(), boxes->data<T>() + boxes->numel(),
+            boxes->data<T>(), clip_func);
+    }
+
+    Eigen::Tensor<T, 2, Eigen::RowMajor> var_et(1, variances.size());
```
Comment (pkuyym) [inline paddle/operators/prior_box_op.h]: I think it is more efficiency to use framework::Tensor whose memory is allocated from internal pool. @qingqing01 Please help to confirm.

Comment (qingqing01) [reply paddle/operators/prior_box_op.h]: Yeah, should use Fluid's memory (or Tenosr) to allocate auxiliary workspace.

File: paddle/operators/prior_box_op.cc [hunk 33]
```diff
@@ -0,0 +1,156 @@
...
+    AddInput("Input",
+             "(Tensor, default Tensor<float>), "
+             "the input feature data of PriorBoxOp, The layout is NCHW.");
+    AddInput("Image",
+             "(Tensor, default Tensor<float>), "
+             "the input image data of PriorBoxOp, The layout is NCHW.");
+    AddOutput("Boxes",
+              "(Tensor, default Tensor<float>), the output prior boxes of "
+              "PriorBoxOp. The layout is [layer_height, layer_width, "
+              "num_priors, 4]. layer_height is the height of input, "
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.cc]: layer_height -> H, 
layer_width -> W

same as below.

File: paddle/operators/prior_box_op.h [hunk 34]
```diff
@@ -0,0 +1,184 @@
...
+
+    T step_w = static_cast<T>(ctx.Attr<float>("step_w"));
+    T step_h = static_cast<T>(ctx.Attr<float>("step_h"));
+    T offset = static_cast<T>(ctx.Attr<float>("offset"));
+
+    auto img_width = image->dims()[3];
+    auto img_height = image->dims()[2];
+
+    auto layer_width = input->dims()[3];
+    auto layer_height = input->dims()[2];
```
Comment (qingqing01) [inline paddle/operators/prior_box_op.h]: layer_width -> feature_width  or width.

Same as layer_height

Comment (wanghaox) [reply paddle/operators/prior_box_op.h]: done

File: paddle/operators/prior_box_op.h (hunk 32 above)
Comment (wanghaox) [reply paddle/operators/prior_box_op.h]: done

File: paddle/operators/prior_box_op.cc (hunk 33 above)
Comment (wanghaox) [reply paddle/operators/prior_box_op.cc]: done

**Pull request merged**